File: day24_part1_failed/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

type ALU struct {
	w, x, y, z int
}

func (a *ALU) execute(instruction string, valueA, valueB int) {
	switch instruction {
	case "add":
		a.add(valueA, valueB)
	case "mul":
		a.mul(valueA, valueB)
	case "div":
		a.div(valueA, valueB)
	case "mod":
		a.mod(valueA, valueB)
	case "eql":
		a.eql(valueA, valueB)
	}
}

func (a *ALU) add(aVar, bVar int) {
	a.setVar(aVar, a.getVar(aVar)+a.getVar(bVar))
}

func (a *ALU) mul(aVar, bVar int) {
	a.setVar(aVar, a.getVar(aVar)*a.getVar(bVar))
}

func (a *ALU) div(aVar, bVar int) {
	a.setVar(aVar, a.getVar(aVar)/a.getVar(bVar))
}

func (a *ALU) mod(aVar, bVar int) {
	a.setVar(aVar, a.getVar(aVar)%a.getVar(bVar))
}

func (a *ALU) eql(aVar, bVar int) {
	if a.getVar(aVar) == a.getVar(bVar) {
		a.setVar(aVar, 1)
	} else {
		a.setVar(aVar, 0)
	}
}

func (a *ALU) setVar(variable int, value int) {
	switch variable {
	case 0:
		a.w = value
	case 1:
		a.x = value
	case 2:
		a.y = value
	case 3:
		a.z = value
	}
}

func (a *ALU) getVar(variable int) int {
	switch variable {
	case 0:
		return a.w
	case 1:
		return a.x
	case 2:
		return a.y
	case 3:
		return a.z
	default:
		return 0
	}
}
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func getVarIdx(varName string) int {
	switch varName {
	case "w":
		return 0
	case "x":
		return 1
	case "y":
		return 2
	case "z":
		return 3
	default:
		return -1
	}
}

func checkNumber(number int, remainingDigits int, program []string, maxModel *int) {
	if remainingDigits == 0 {
		alu := ALU{0, 0, 0, 0}
		for _, line := range program {
			parts := strings.Split(line, " ")
			switch parts[0] {
			case "inp":
				alu.setVar(getVarIdx(parts[1]), number%10)
				number /= 10
			default:
				valueA := getVarIdx(parts[1])
				valueB, _ := strconv.Atoi(parts[2])
				if valueB < 0 || valueB > 3 {
					alu.execute(parts[0], valueA, valueB)
				} else {
					alu.execute(parts[0], valueA, getVarIdx(parts[2]))
				}
			}
		}

		if alu.z == 0 {
			*maxModel = max(*maxModel, number)
		}
		return
	}

	for i := 1; i <= 9; i++ {
		checkNumber(number*10+i, remainingDigits-1, program, maxModel)
	}
}

func main() {
	data, err := ioutil.ReadFile("day24_part1/test.txt")
	if err != nil {
		panic(err)
	}

	lines := strings.Split(string(data), "\n")

	maxModel := 0
	checkNumber(0, 14, lines, &maxModel)
	fmt.Println("The largest valid model number accepted by MONAD:", maxModel)
}

File: day24_part1_failed/main.py
def read_instructions(file_path):
    with open(file_path, 'r') as f:
        return [line.strip().split() for line in f.readlines()]

def execute_instructions(instructions, inputs):
    variables = {'w': 0, 'x': 0, 'y': 0, 'z': 0}
    input_idx = 0

    for instr in instructions:
        op, a, *rest = instr
        b = rest[0] if rest else None

        if op == "inp":
            variables[a] = inputs[input_idx]
            input_idx += 1
        elif op == "add":
            variables[a] += int(b) if b.isdigit() else variables[b]
        elif op == "mul":
            variables[a] *= int(b) if b.isdigit() else variables[b]
        elif op == "div":
            divisor = int(b) if b.isdigit() else variables[b]
            if divisor != 0:
                variables[a] = variables[a] // divisor
        elif op == "mod":
            divisor = int(b) if b.isdigit() else variables[b]
            if 0 < divisor:
                variables[a] %= divisor
        elif op == "eql":
            variables[a] = 1 if variables[a] == (int(b) if b.isdigit() else variables[b]) else 0

    return variables

def find_largest_model_number(file_path):
    instructions = read_instructions(file_path)
    largest_number = 0

    def search(number, depth):
        nonlocal largest_number
        if depth == 14:
            variables = execute_instructions(instructions, [int(d) for d in str(number)])
            if variables['z'] == 0:
                largest_number = max(largest_number, number)
        else:
            for digit in range(1, 10):
                search(number * 10 + digit, depth + 1)

    search(0, 0)
    return largest_number

if __name__ == "__main__":
    file_path = "./test.txt"
    print(find_largest_model_number(file_path))
